presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkMapFlatWriter.java

```diff
@@ -123,64 +129,125 @@ private void doWrite(BenchmarkData data)
+                int keyCount = (int) (distinctKeyCount * weights.get(i) / weightSum);
+                for (int j = 0; j < keyCount; j++) {
+                    keyFrequencies[j + offset] = frequency;
+                }
+                offset += keyCount;
+            }
 
             for (int i = 0; i < blockCount; i++) {
-                MapBlockBuilder mapBlockBuilder = (MapBlockBuilder) type.createBlockBuilder(null, positionsPerBlock);
-                for (int j = 0; j < positionsPerBlock; j++) {
-                    Collections.shuffle(keys, random);
-                    int entryKeyCount = random.nextInt(maxKeysPerEntry);
+                MapBlockBuilder mapBlockBuilder = (MapBlockBuilder) this.type.createBlockBuilder(null, positionsPerBlock);
+                BlockBuilder mapKeyBuilder = mapBlockBuilder.getKeyBlockBuilder();
+                BlockBuilder mapValueBuilder = mapBlockBuilder.getValueBlockBuilder();
 
+                for (int j = 0; j < positionsPerBlock; j++) {
                     mapBlockBuilder.beginDirectEntry();
-                    BlockBuilder keyBuilder = mapBlockBuilder.getKeyBlockBuilder();
-                    BlockBuilder valueBuilder = mapBlockBuilder.getValueBlockBuilder();
 
-                    for (int k = 0; k < entryKeyCount; k++) {
-                        keyType.writeLong(keyBuilder, keys.get(k));
-                        keyType.writeLong(valueBuilder, keys.get(k) * 2L + 1);
+                    int entryKeyCount = 0;
+
+                    // add keys
+                    for (int k = 0; k < distinctKeyCount; k++) {
+                        if (random.nextDouble() < keyFrequencies[k]) {
+                            keyType.writeLong(mapKeyBuilder, keys.get(k));
+                            entryKeyCount++;
+                        }
+                    }
+
+                    // add values
+                    if (valueType == BigintType.BIGINT) {
+                        for (int k = 0; k < entryKeyCount; k++) {
+                            valueType.writeLong(mapValueBuilder, random.nextLong());
+                        }
+                    }
+                    else if (valueType instanceof ArrayType) {
+                        ArrayType arrayValueType = (ArrayType) valueType;
+                        Type elementType = arrayValueType.getElementType();
+                        ArrayBlockBuilder arrayBlockBuilder = (ArrayBlockBuilder) mapValueBuilder;
+                        BlockBuilder arrayElementValueBuilder = arrayBlockBuilder.getElementBlockBuilder();
+                        for (int k = 0; k < entryKeyCount; k++) {
+                            arrayBlockBuilder.beginDirectEntry();
+                            for (int c = 0; c < childElementValueCount; c++) {
+                                elementType.writeLong(arrayElementValueBuilder, random.nextLong());
+                            }
+                            mapValueBuilder.closeEntry();
```

Comment:
openEntry is on arrayBuilder but close is on mapValueBuilder, this looks confusing to me.